### DISCUSSION TITLE
Possible to remove elements in an array?

### DISCUSSION THREAD

**Discussion by ewmailing** [pallene-lang/pallene#503] [Q&A] [answered]
Is it possible to remove elements in an array in Pallene right now?
I tried the easiest case of removing the last element in an array by assigning it nil. But I think the type system is rejecting that. (I haven't gotten to the hard case of removing an element in the middle of an array.)

I've been working around it for the moment, where I just create a new array, copy over all the elements I want to keep, and let the old array be collected. But I'm reaching a point in my experiment program where potentially multiple things are keeping a reference to the same array, so trying to swap it out like that doesn't work so well.

**Comment by hugomg** [comment, marked as answer]
I believe the the only way to do that right now may be to cast the array to type `{any}`.

**Reply by ewmailing** [reply]
Thanks. That seemed to do the trick.

**Reply by ewmailing** [reply]
Out of curiosity, when I want to shrink an array, do I need to nil from the end of the array and work backwards, or can I start from the middle and go to the end?

**Reply by srijan-paul** [reply]
Could you explain?
If you nil out a middle element, `ipairs` won't go over the entire array.

**Reply by ewmailing** [reply]
I know holes in an array will screw up things like ipairs. However, if the array started contiguous, and holes were only transitory in clearing to the end of the array, does that break anything?
I'm using numeric for. So is there a difference between the following two examples? Does the first solution break things?

```
-- For simplicity, assume the array has 10 elements, and I want to delete the last 5.
for i=6, 10 do
    array[i] = nil
end
```
vs.
```
for i=10,6, -1 do
    array[i] = nil
end
```

**Reply by hugomg** [reply]
I wouldn't expect any logical errors because this numeric for loop is not using `#` or `ipairs` and therefore does not care if the array has holes or not. If there is any difference between the two versions, I would expect to be a performance difference. Under the hood, Lua implements tables using an "array" part and a "hash" part and the heuristic likes when arrays won't have holes.

I don't believe this is actually a problem in practice though. If I remember correctly, Lua only grows or shrinks the memory buffer when it is  inserting new elements into the table, not when deleting them. If I am remembering things correctly then there should be no difference between deleting forwards or backwards. However, if I'm remembering incorrectly then iterating backwards is better. I suppose it wouldn't hurt to iterate backwards then.
